app/test/vesting_test.go

```diff
@@ -0,0 +1,632 @@
+	default:
+		return "unknown"
+	}
+}
+
+type VestingModuleTestSuite struct {
+	suite.Suite
+
+	// accounts is a map from accountType to accountDispenser
+	accounts    sync.Map
+	accountsMut sync.Mutex
+
+	kr   keyring.Keyring
+	cctx testnode.Context
+	ecfg encoding.Config
+}
+
+func TestVestingModule(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping vesting module test suite in short mode.")
+	}
+	suite.Run(t, new(VestingModuleTestSuite))
+}
+
+func (s *VestingModuleTestSuite) SetupSuite() {
+	t := s.T()
+	t.Log("setting up vesting module test suite")
+
+	s.kr = testfactory.GenerateKeyring()
+
+	genOpts := []testnode.GenesisOption{}
+	genOpts = append(genOpts, s.initRegularAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initDelayedVestingAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initPeriodicVestingAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initContinuousVestingAccounts(totalAccountsPerType))
+
+	s.startNewNetworkWithGenesisOpt(genOpts...)
+}
+
+// startNewNetworkWithGenesisOpt creates a new test network with the specified genesis options for the VestingModuleTestSuite.
+// It initializes a default Tendermint configuration (tmCfg) and default application configuration (appConf).
+// The target block time is set to 1 millisecond. It applies the given genesis options to the test network
+// and stores the created client context (cctx) in the VestingModuleTestSuite.
+// The keyring of the context is set to the keyring (s.kr) of the VestingModuleTestSuite.
+func (s *VestingModuleTestSuite) startNewNetworkWithGenesisOpt(genesisOpts ...testnode.GenesisOption) {
+	s.ecfg = encoding.MakeConfig(app.ModuleEncodingRegisters...)
+	cfg := testnode.DefaultConfig().WithGenesisOptions(testnode.ImmediateProposals(s.ecfg.Codec))
+	cfg.GenesisOptions = genesisOpts
+
+	cctx, _, _ := testnode.NewNetwork(s.T(), cfg)
+	s.cctx = cctx
+	s.cctx.Keyring = s.kr
+}
+
+func (s *VestingModuleTestSuite) TestGenesisDelayedVestingAccountsTransferLocked() {
+	require.NoError(s.T(), s.cctx.WaitForNextBlock())
+
+	// find and test a vesting account with endTime which is
+	// at least 10 seconds away from now to give the tx enough time to complete
+	_, name, err := s.getAnUnusedDelayedVestingAccount(tmtime.Now().Unix() + vestingDelayPerTx)
+	require.NoError(s.T(), err)
+
+	s.testTransferMustFail(name, vestingAmount)
+}
+
+func (s *VestingModuleTestSuite) TestGenesisDelayedVestingAccountsTransferUnLocked() {
+	require.NoError(s.T(), s.cctx.WaitForNextBlock())
+
+	// find and test a vesting account with endTime which is already passed
+	vAcc, name, err := s.getAnUnusedDelayedVestingAccount(0)
+	require.NoError(s.T(), err)
+
+	// It is possible in some cases the given account is not unlocked
+	// so if that's the case, we wait a bit here
+	for vAcc.GetVestedCoins(tmtime.Now()).IsZero() {
+		time.Sleep(time.Second)
+	}
+	minExpectedSpendableBal := vAcc.GetVestedCoins(tmtime.Now()).AmountOf(app.BondDenom).Int64()
```

Comment:
[optional]
The `GetVestedCoins` returns nil if no coins are vested. Do you think it makes sense to split this into:
```
	minExpectedSpendableBal := vAcc.GetVestedCoins(tmtime.Now())
	require.NotNil(s.T(), minExpectedSpendableBal)
	require.NoError(s.T(), s.cctx.WaitForNextBlock())

	s.testTransferMustSucceed(name, minExpectedSpendableBal.AmountOf(app.BondDenom).Int64())
```
And anywhere this is used?

Reply:
Thanks for pointing it out. Yes we can add it for the unlocked balances tests. will add it.

Reply:
Thanks for pointing it out. Yes we can add it for the unlocked balances tests. will add it.

Reply:
done